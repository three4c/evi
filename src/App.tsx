import { useEffect, useRef } from "react";
import * as stylex from "@stylexjs/stylex";
import "./reset.css";

const styles = stylex.create({
  div: {
    backdropFilter: "blur()12px",
    backgroundColor: "rgba(255, 255, 255, 0.65)",
    fontSize: "2em",
    display: "flex",
    alignItems: "center",
    justifyContent: "center",
    width: 48,
    height: 48,
  },
});

const DOM_ARRAY = ["INPUT", "TEXTAREA"];

const App = () => {
  const isStartVim = useRef(false);

  const startVim = (element: Element, e: KeyboardEvent) => {
    if (
      (element instanceof HTMLInputElement ||
        element instanceof HTMLTextAreaElement) &&
      element.selectionStart !== null
    ) {
      let start = element.selectionStart;
      let end = start + 1;

      if (e.key === "h" && start) {
        start--;
        end = start + 1;
      }

      if (e.key === "l" && start !== element.value.length - 1) {
        start++;
        end = start + 1;
      }

      let charCount = 0;
      let currentLine = 0;
      let col = 0;
      const lines = element.value.split("\n");

      for (let i = 0; i < lines.length; i++) {
        const lineLength = lines[i].length + 1;
        if (charCount + lineLength > start) {
          currentLine = i;
          col = start - charCount;
          break;
        }
        charCount += lineLength;
      }

      if (e.key === "j" && currentLine + 1 < lines.length) {
        const next = charCount + lines[currentLine].length + 1;
        start = next + col;
        end = start + 1;
      }

      if (e.key === "k" && currentLine > 0) {
        const prev = charCount - (lines[currentLine - 1].length + 1);
        start = prev + col;
        end = start + 1;
      }

      element.setSelectionRange(start, end);

      if (e.key === "i") {
        element.setSelectionRange(start, start);
        isStartVim.current = false;
      }
    }
  };

  useEffect(() => {
    const onKeyDown = (e: KeyboardEvent) => {
      const activeElement = document.activeElement;

      if (!activeElement || !DOM_ARRAY.includes(activeElement.tagName)) {
        return;
      }

      if (e.ctrlKey && e.key === "Enter") {
        isStartVim.current = true;
      }

      if (e.ctrlKey && e.key === "q") {
        isStartVim.current = false;
      }

      if (isStartVim.current) {
        startVim(activeElement, e);
        e.preventDefault();
      }
    };

    window.addEventListener("keydown", onKeyDown);
    return () => window.removeEventListener("keydown", onKeyDown);
  }, []);

  return (
    <>
      <div {...stylex.props(styles.div)}>ü¶ê</div>
      <input type="text" />
      <textarea />
    </>
  );
};

export default App;
