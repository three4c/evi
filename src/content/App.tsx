import { useEffect, useRef, useState } from "react";
import {
  loadShortcuts,
  matchesShortcut,
  onShortcutsChanged,
  ShortcutConfig,
} from "../utils/shortcuts";

type MODE_TYPE = "normal" | "insert" | "visual";

const DOM_ARRAY = ["INPUT", "TEXTAREA"];

const App: React.FC = () => {
  const mode = useRef<MODE_TYPE>("insert");
  const pos = useRef({
    start: 0,
    end: 0,
  });
  const originalPos = useRef<
    | {
        start: number;
        end: number;
        currentLine: number;
      }
    | undefined
  >(undefined);
  const [shortcuts, setShortcuts] = useState<Record<string, ShortcutConfig>>(
    {},
  );

  const getElement = (element: Element | null) =>
    element instanceof HTMLInputElement ||
    element instanceof HTMLTextAreaElement
      ? element
      : null;

  const getLines = (
    element: HTMLInputElement | HTMLTextAreaElement,
    start: number,
  ) => {
    let charCount = 0;
    let currentLine = 0;
    let col = 0;
    const lines = element.value.split("\n");

    for (let i = 0; i < lines.length; i++) {
      const linesLength = lines[i].length + 1;

      if (charCount + linesLength > start) {
        currentLine = i;
        col = start - charCount;
        break;
      }

      charCount += linesLength;
    }

    return {
      lines,
      charCount,
      currentLine,
      col,
    };
  };

  const startVim = async (
    element: HTMLInputElement | HTMLTextAreaElement,
    e: KeyboardEvent,
  ) => {
    let { start, end } = pos.current;
    const { length } = element.value;

    if (mode.current === "normal") {
      start = element.selectionStart || 0;
      end = start + 1;
    }

    const { lines, charCount, currentLine, col } = getLines(element, start);
    const { currentLine: endCurrentLine } = getLines(element, end);

    if (
      mode.current === "normal" &&
      !matchesShortcut(e, shortcuts.normal_mode)
    ) {
      if (e.key === "h" && col) {
        start--;
        end = start + 1;
      }

      if (e.key === "l" && col !== lines[currentLine].length - 1) {
        start++;
        end = start + 1;
      }

      if (e.key === "j" && currentLine + 1 < lines.length) {
        const nextLineLength = lines[currentLine + 1].length;
        const next = charCount + lines[currentLine].length + 1;

        if (nextLineLength) {
          const overCharCount =
            col >= nextLineLength ? col - nextLineLength + 1 : 0;
          start = next + col - overCharCount;
        } else {
          start = next;
        }
        end = start + 1;
      }

      if (e.key === "k" && currentLine > 0) {
        const prevLineLength = lines[currentLine - 1].length;
        const prev = charCount - (lines[currentLine - 1].length + 1);

        if (prevLineLength) {
          const overCharCount =
            col >= prevLineLength ? col - prevLineLength + 1 : 0;
          start = prev + col - overCharCount;
        } else {
          start = prev;
        }
        end = start + 1;
      }

      if (e.key === "J") {
        const nextBreak = element.value.indexOf("\n", start);
        if (nextBreak >= 0) {
          start = nextBreak;
          end = nextBreak + 1;
          element.value = [
            element.value.slice(0, start),
            element.value.slice(end, length),
          ].join(" ");
        }
      }

      if (e.key === "o" && element.tagName === "TEXTAREA") {
        const nextBreak = element.value.indexOf("\n", start);
        start = nextBreak === -1 ? length : nextBreak;
        element.value = [
          element.value.slice(0, start),
          element.value.slice(start, length),
        ].join("\n");
        start = end = start + 1;
        mode.current = "insert";
      }

      if (e.key === "O" && element.tagName === "TEXTAREA") {
        const prevBreak = element.value.lastIndexOf("\n", start - 1);
        start = prevBreak === -1 ? 0 : prevBreak;
        element.value = [
          element.value.slice(0, start),
          element.value.slice(start, length),
        ].join("\n");
        start = end = start + (start ? 1 : 0);
        mode.current = "insert";
      }

      if (e.key === "p") {
        const text = await navigator.clipboard.readText();
        if (text) {
          const pos = lines[currentLine].length === 0 ? start : end;
          element.value = [
            element.value.slice(0, pos),
            element.value.slice(pos, length),
          ].join(text);
          start = start + text.length;
          end = start + 1;
        }
      }

      if (e.key === "P") {
        const text = await navigator.clipboard.readText();
        if (text) {
          element.value = [
            element.value.slice(0, start),
            element.value.slice(start, length),
          ].join(text);
          start = start + text.length - 1;
          end = start + 1;
        }
      }

      if (lines[currentLine].length && col === lines[currentLine].length) {
        start = start - 1;
        end = start + 1;
      }
    }

    if (mode.current === "visual" && originalPos.current) {
      const {
        start: oStart,
        end: oEnd,
        currentLine: oCurrentLine,
      } = originalPos.current;

      const isSingleChar = end - start === 1;
      const shouldMoveStart = isSingleChar ? start <= oStart : start < oStart;
      const shouldMoveEnd = isSingleChar ? end >= oEnd : end > oEnd;
      const atLeftEdge = start === 0 && oStart === 0;
      const atRightEdge = end === length && oEnd === length;

      const currentLineLength = lines[currentLine].length + 1;
      const endCurrentLineLength = lines[endCurrentLine].length + 1;
      const clampToTextBounds = (value: number) =>
        Math.max(0, Math.min(value, length));

      if (e.key === "h") {
        if (start > 0) {
          if (shouldMoveStart) {
            start--;
          } else {
            end--;
          }
        } else if (atLeftEdge && end !== 1 && !isSingleChar) {
          end--;
        }
      }

      if (e.key === "l") {
        if (end < length) {
          if (shouldMoveEnd) {
            end++;
          } else {
            start++;
          }
        } else if (atRightEdge && start !== length - 1 && !isSingleChar) {
          start++;
        }
      }

      if (e.key === "j") {
        const canMoveDown = currentLine + 1 < lines.length;
        const expandDown = currentLine >= oCurrentLine;
        if (canMoveDown && expandDown) {
          end = clampToTextBounds(end + currentLineLength);
        } else {
          start = start + currentLineLength;
          if (start > end) {
            start = oStart;
            end = length;
          }
        }
      }

      if (e.key === "k") {
        const canMoveUp = endCurrentLine > 0;
        const expandUp = endCurrentLine <= oCurrentLine;
        if (canMoveUp && expandUp) {
          start = clampToTextBounds(start - endCurrentLineLength);
        } else {
          end = end - endCurrentLineLength;
          if (end < start) {
            start = 0;
            end = oEnd;
          }
        }
      }

      if (e.key === "p" || e.key === "P") {
        const text = await navigator.clipboard.readText();
        if (text) {
          element.setRangeText(text);
          start = start + text.length - 1;
          end = start + 1;
          mode.current = "normal";
        }
      }

      if (e.key === "y") {
        const text = window.getSelection()?.toString();
        if (text) {
          navigator.clipboard.writeText(text);
          mode.current = "normal";
          start = oStart;
          end = oEnd;
        }
      }
    }

    if (e.key === "i") {
      end = start;
    }

    if (e.key === "I") {
      start = end = element.value.lastIndexOf("\n", start) + 1;
    }

    if (e.key === "a") {
      start = end;
    }

    if (e.key === "A") {
      start = end = element.value.indexOf("\n", start);
    }

    if (["i", "I", "a", "A"].includes(e.key)) {
      mode.current = "insert";
    }

    if (e.key === "v") {
      if (mode.current === "normal") {
        mode.current = "visual";
        originalPos.current = { start, end, currentLine };
      } else {
        mode.current = "normal";
        start = originalPos.current?.start || start;
        end = originalPos.current?.end || end;
        originalPos.current = undefined;
      }
    }

    pos.current = { start, end };
    element.setSelectionRange(start, end);
  };

  useEffect(() => {
    loadShortcuts().then(setShortcuts);
    onShortcutsChanged(setShortcuts);
  }, []);

  useEffect(() => {
    const onKeyDown = (e: KeyboardEvent) => {
      const activeElement = document.activeElement;
      const element = getElement(activeElement);

      if (!element) {
        return;
      }

      if (!DOM_ARRAY.includes(element.tagName)) {
        return;
      }

      if (shortcuts.normal_mode && matchesShortcut(e, shortcuts.normal_mode)) {
        mode.current = "normal";
      }

      if (["normal", "visual"].includes(mode.current)) {
        startVim(element, e);
        e.preventDefault();
      }
    };

    window.addEventListener("keydown", onKeyDown);
    return () => {
      window.removeEventListener("keydown", onKeyDown);
    };
  }, [shortcuts]);

  return null;
};

export default App;
