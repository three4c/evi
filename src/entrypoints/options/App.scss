.App {
  font-family: system-ui;
  font-size: 14px;
  padding: 20px;

  &__section {
    margin-bottom: 20px;
  }

  &__subtitle {
    margin-bottom: 10px;
    font-size: 18px;
    font-weight: bold;
  }

  &__input-group {
    display: flex;
    align-items: center;
    gap: 10px;
  }

  &__input {
    padding: 6px 12px;
    border: 1px solid #ccc;
    border-radius: 4px;
    min-width: 150px;
    background-color: #fff;

    &--editing {
      background-color: #f0f8ff;
    }
  }

  &__button {
    padding: 6px 12px;
    border: 1px solid #ccc;
    border-radius: 4px;
    background-color: #fff;
    cursor: pointer;
  }

  &__message {
    padding: 10px;
    border-radius: 4px;

    &--error {
      background-color: #ffebee;
      border: 1px solid #ffcdd2;
      color: #c62828;
    }

    &--success {
      background-color: #e8f5e8;
      border: 1px solid #c8e6c9;
      color: #2e7d32;
    }
  }
}
