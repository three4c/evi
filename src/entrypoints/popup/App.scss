.App {
  background: linear-gradient(180deg, rgba(255, 175, 0, 1) 0%, rgba(255, 34, 101, 1) 100%);
  display: inline-flex;
  align-items: center;
  justify-content: center;
  padding: 8px 16px;
  user-select: none;

  &__text {
    color: #fff;
    font-weight: 600;
    font-family: 'Rubik', serif;
    font-size: 16px;
    pointer-events: none;
  }

  &__icon {
    margin-right: 8px;
    pointer-events: none;
  }

  &__settings {
    margin-left: 8px;
    padding: 0;
    background: transparent;
    border: none;

    &:hover {
      cursor: pointer;
      filter: brightness(0.8);
    }
  }

  &__child {
    width: 100%;
  }

  &__button {
    margin-top: 8px;
  }

  &--dev {
    display: flex;
    flex-direction: column;
    justify-content: center;
    align-items: center;
    gap: 16px;
    width: 300px;
    padding: 16px;
  }
}
