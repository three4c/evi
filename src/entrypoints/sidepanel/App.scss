.App {
  font-family: system-ui;
  font-size: 14px;
  padding: 16px 16px 56px;

  &__section {
    margin-bottom: 32px;
  }

  &__subtitle {
    margin-bottom: 16px;
    font-size: 18px;
    font-weight: bold;
  }

  &__inputGroup {
    display: flex;
    align-items: center;
    gap: 8px;
    margin-bottom: 8px;

    > * {
      flex-shrink: 0;
    }
  }

  &__label {
    width: 130px;
    text-transform: capitalize;
  }

  &__input {
    padding: 8px 12px;
    border: 1px solid #ccc;
    border-radius: 4px;
    width: 100px;
    background-color: #fff;

    &--editing {
      background-color: #f0f8ff;
    }
  }

  &__button {
    padding: 8px 12px;
    border: 1px solid #ccc;
    border-radius: 4px;
    background-color: #fff;
    cursor: pointer;

    &--reset {
      position: fixed;
      left: 16px;
      bottom: 16px;
    }
  }

  &__buttonGroup {
    display: flex;
    gap: 4px;
  }

  &__message {
    position: fixed;
    right: 0;
    bottom: 0;
    margin: 16px;
    padding: 8px;
    border-radius: 4px;

    &--error {
      background-color: #ffebee;
      border: 1px solid #ffcdd2;
      color: #c62828;
    }

    &--success {
      background-color: #e8f5e8;
      border: 1px solid #c8e6c9;
      color: #2e7d32;
    }
  }
}
