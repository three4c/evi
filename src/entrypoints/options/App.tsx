import { useEffect, useState } from "react";
import {
  DEFAULT_SHORTCUTS,
  getShortcutString,
  loadShortcuts,
  parseKeyboardEvent,
  type ShortcutConfig,
  saveShortcuts,
  validateShortcut,
} from "../../utils/shortcuts";
import "./App.scss";

const App: React.FC = () => {
  const [shortcuts, setShortcuts] =
    useState<Record<string, ShortcutConfig>>(DEFAULT_SHORTCUTS);
  const [isEditing, setIsEditing] = useState(false);
  const [message, setMessage] = useState("");

  useEffect(() => {
    loadShortcuts().then(setShortcuts);
  }, []);

  const handleEdit = () => {
    setIsEditing(true);
    setMessage("キーの組み合わせを押してください...");
  };

  const handleCancel = () => {
    setIsEditing(false);
    setMessage("");
  };

  useEffect(() => {
    if (!isEditing) {
      return;
    }

    const handleKeyDown = (e: KeyboardEvent) => {
      if (!isEditing) return;

      e.preventDefault();

      const newShortcut = parseKeyboardEvent(e);

      const modifierKeys = [
        "control",
        "shift",
        "alt",
        "meta",
        "cmd",
        "command",
      ];
      if (modifierKeys.includes(newShortcut.key.toLowerCase())) {
        setMessage("追加のキーを押してください...");
        return;
      }

      const error = validateShortcut(newShortcut);
      if (error) {
        setMessage(`エラー: ${error}`);
        return;
      }

      const updatedShortcuts = { ...shortcuts, normal_mode: newShortcut };
      setShortcuts(updatedShortcuts);
      setIsEditing(false);

      saveShortcuts(updatedShortcuts).then(() => {
        setMessage("保存完了！");
        setTimeout(() => setMessage(""), 3000);
      });
    };

    document.addEventListener("keydown", handleKeyDown);
    return () => document.removeEventListener("keydown", handleKeyDown);
  }, [isEditing, shortcuts]);

  return (
    <div className="App">
      <div className="App__section">
        <h3 className="App__subtitle">Normal Mode</h3>
        <div className="App__input-group">
          <input
            type="text"
            value={
              isEditing
                ? "入力待機中..."
                : getShortcutString(shortcuts.normal_mode)
            }
            readOnly
            className={`App__input ${isEditing ? "App__input--editing" : ""}`}
          />
          <button
            type="button"
            className="App__button"
            onClick={isEditing ? handleCancel : handleEdit}
          >
            {isEditing ? "キャンセル" : "編集"}
          </button>
        </div>
      </div>

      {message && (
        <div
          className={`App__message ${
            message.includes("エラー")
              ? "App__message--error"
              : "App__message--success"
          }`}
        >
          {message}
        </div>
      )}
    </div>
  );
};

export default App;
