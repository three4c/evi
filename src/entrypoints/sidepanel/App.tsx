import { useCallback, useEffect, useState } from "react";
import { DEFAULT_KEYMAPS } from "@/keymaps";
import type { Keymap, Keymaps, MODE_TYPE } from "@/utils";
import {
  detectModifierKey,
  loadKeymaps,
  markKeymapsUpdating,
  resetKeymaps,
  saveKeymaps,
  sendKeymaps,
} from "@/utils";
import "./App.scss";

type ALL_MODE_TYPE = MODE_TYPE | "common";
const MESSAGE_DISPLAY_TIME = 3000;

const App: React.FC = () => {
  const [keymaps, setKeymaps] = useState<Keymaps>(DEFAULT_KEYMAPS);
  const [isEditing, setIsEditing] = useState(false);
  const [editingMode, setEditingMode] = useState<ALL_MODE_TYPE | null>(null);
  const [editingCommand, setEditingCommand] = useState<string | null>(null);
  const [message, setMessage] = useState("");
  const [currentKeySequence, setCurrentKeySequence] = useState<string[]>([]);
  const [validationError, setValidationError] = useState("");

  const validateKeySequence = useCallback(
    (sequence: string[], mode: ALL_MODE_TYPE, command: string) => {
      if (sequence.length === 0) {
        return "キーシーケンスが空です";
      }

      const allKeymapsForMode =
        mode === "common"
          ? {
              ...keymaps.common,
              ...keymaps.normal,
              ...keymaps.visual,
            }
          : { ...keymaps[mode], ...keymaps.common };
      const newKeyString = sequence.join(" ");

      // 重複チェック（現在編集中のコマンド以外）
      for (const [cmd, keyStr] of Object.entries(allKeymapsForMode)) {
        if (cmd !== command && keyStr === newKeyString) {
          return `キー "${newKeyString}" は既に "${cmd}" で使用されています`;
        }
      }

      // プレフィックス競合チェック
      for (const [cmd, keyStr] of Object.entries(allKeymapsForMode)) {
        if (cmd === command) continue;

        const existingKeys = keyStr.split(" ");

        // 既存の1キーマップが新しいシーケンスの最初のキーと同じ場合
        if (
          existingKeys.length === 1 &&
          sequence.length > 1 &&
          existingKeys[0] === sequence[0]
        ) {
          return `キー "${existingKeys[0]}" が既に "${cmd}" で使用されているため、"${newKeyString}" は設定できません`;
        }

        // 新しいシーケンスが1キーで、既存のマルチキーシーケンスの最初のキーと同じ場合
        if (
          sequence.length === 1 &&
          existingKeys.length > 1 &&
          sequence[0] === existingKeys[0]
        ) {
          return `キー "${sequence[0]}" は "${cmd}" のキーシーケンス "${keyStr}" と競合します`;
        }
      }

      return "";
    },
    [keymaps],
  );

  const handleEdit = (mode: ALL_MODE_TYPE, command: string) => {
    setIsEditing(true);
    setEditingMode(mode);
    setEditingCommand(command);
    setCurrentKeySequence([]);
    setValidationError("");
    setMessage("新しいキーを押してください...");
    setTimeout(() =>
      document.querySelector<HTMLInputElement>(".App__input--editing")?.focus(),
    );
  };

  const handleCancel = () => {
    setIsEditing(false);
    setEditingMode(null);
    setEditingCommand(null);
    setCurrentKeySequence([]);
    setValidationError("");
    setMessage("");
  };

  const handleConfirm = async () => {
    if (!editingMode || !editingCommand || currentKeySequence.length === 0)
      return;

    const error = validateKeySequence(
      currentKeySequence,
      editingMode,
      editingCommand,
    );
    if (error) {
      setValidationError(error);
      return;
    }

    const newKeyString = currentKeySequence.join(" ");
    const updatedKeymaps = { ...keymaps };
    updatedKeymaps[editingMode][editingCommand] = newKeyString;

    setKeymaps(updatedKeymaps);
    setIsEditing(false);
    setEditingMode(null);
    setEditingCommand(null);
    setCurrentKeySequence([]);
    setValidationError("");

    await saveKeymaps(updatedKeymaps);
    await sendKeymaps(updatedKeymaps);

    setMessage("保存完了！");
    setTimeout(() => setMessage(""), MESSAGE_DISPLAY_TIME);
  };

  const updateKeymaps = useCallback((savedKeymaps: Keymaps) => {
    setKeymaps({
      common: { ...DEFAULT_KEYMAPS.common, ...savedKeymaps.common },
      insert: { ...DEFAULT_KEYMAPS.insert, ...savedKeymaps.insert },
      normal: { ...DEFAULT_KEYMAPS.normal, ...savedKeymaps.normal },
      visual: { ...DEFAULT_KEYMAPS.visual, ...savedKeymaps.visual },
    });
  }, []);

  const handleReset = async () => {
    setKeymaps(DEFAULT_KEYMAPS);
    await resetKeymaps();
    await loadKeymaps().then(updateKeymaps);
    setMessage("キーマップをデフォルトにリセットしました！");
    setTimeout(() => setMessage(""), MESSAGE_DISPLAY_TIME);
  };

  useEffect(() => {
    loadKeymaps().then(updateKeymaps);
    markKeymapsUpdating();
  }, [updateKeymaps]);

  useEffect(() => {
    if (!isEditing) {
      return;
    }

    const handleKeyDown = (e: KeyboardEvent) => {
      if (!isEditing || !editingMode || !editingCommand) return;

      // キーボードで確定やキャンセルにフォーカスできるようにする
      if (["tab", "enter"].includes(e.key.toLowerCase())) {
        return;
      }

      e.preventDefault();

      const key = detectModifierKey(e);

      const modifierKeys = ["ctrl", "alt", "meta", "cmd", "command"];
      if (modifierKeys.includes(e.key.toLowerCase())) {
        setMessage("追加のキーを押してください...");
        return;
      }

      if (e.key.length !== 1) {
        setMessage("文字キーを入力してください");
        return;
      }

      const newSequence = [...currentKeySequence, key];
      setCurrentKeySequence(newSequence);

      const error = validateKeySequence(
        newSequence,
        editingMode,
        editingCommand,
      );
      setValidationError(error);

      if (error) {
        setMessage(`エラー: ${error}`);
      } else {
        setMessage(
          `キーシーケンス: ${newSequence.join(" ")} - 確定またはキャンセルしてください`,
        );
      }
    };

    document.addEventListener("keydown", handleKeyDown);
    return () => document.removeEventListener("keydown", handleKeyDown);
  }, [
    isEditing,
    editingMode,
    editingCommand,
    currentKeySequence,
    validateKeySequence,
  ]);

  const renderKeymapSection = (
    title: string,
    mode: ALL_MODE_TYPE,
    keymap: Keymap,
  ) => (
    <div className="App__section">
      <h3 className="App__subtitle">{title}</h3>
      {Object.entries(keymap).map(([command, key]) => (
        <div key={command} className="App__inputGroup">
          <label className="App__label" htmlFor={`App__input--${command}`}>
            {command.replaceAll("_", " ")}:
          </label>
          <input
            id={`App__input--${command}`}
            type="text"
            value={
              isEditing && editingMode === mode && editingCommand === command
                ? currentKeySequence.join(" ") || "待機中..."
                : key
            }
            readOnly
            className={`App__input ${
              isEditing && editingMode === mode && editingCommand === command
                ? "App__input--editing"
                : ""
            } ${
              validationError &&
              editingMode === mode &&
              editingCommand === command
                ? "App__input--error"
                : ""
            }`}
          />
          {isEditing && editingMode === mode && editingCommand === command ? (
            <div className="App__buttonGroup">
              <button
                type="button"
                className="App__button App__button--confirm"
                onClick={handleConfirm}
                disabled={currentKeySequence.length === 0 || !!validationError}
              >
                確定
              </button>
              <button
                type="button"
                className="App__button App__button--cancel"
                onClick={handleCancel}
              >
                キャンセル
              </button>
            </div>
          ) : (
            <button
              type="button"
              className="App__button"
              onClick={() => handleEdit(mode, command)}
              disabled={isEditing}
            >
              編集
            </button>
          )}
        </div>
      ))}
    </div>
  );

  return (
    <div className="App">
      {renderKeymapSection("Insert Mode Keymaps", "insert", keymaps.insert)}
      {renderKeymapSection("Normal Mode Keymaps", "normal", keymaps.normal)}
      {renderKeymapSection("Visual Mode Keymaps", "visual", keymaps.visual)}
      {renderKeymapSection("Normal / Visual Keymaps", "common", keymaps.common)}

      <button
        type="button"
        className="App__button App__button--reset"
        onClick={handleReset}
        disabled={isEditing}
      >
        デフォルトにリセット
      </button>

      {message && (
        <div
          className={`App__message ${
            message.includes("エラー")
              ? "App__message--error"
              : "App__message--success"
          }`}
        >
          {message}
        </div>
      )}
    </div>
  );
};

export default App;
