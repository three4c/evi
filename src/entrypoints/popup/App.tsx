import { useEffect } from "react";
import { initVim } from "../content";
import "./App.scss";

const App: React.FC = () => {
  const isDev = import.meta.env.DEV;

  const openShortcutSettings = () => chrome.runtime.openOptionsPage();
  const testString = "aaaaa\nbbbbb\nccccc";

  useEffect(() => {
    if (isDev) initVim();
  }, []);

  return isDev ? (
    <div className="App App--dev">
      <p className="App__text">Vite Dev Mode</p>
      <input className="App__child" type="text" />
      <textarea className="App__child" defaultValue={testString} />
      <button
        type="button"
        className="App__child App__button"
        onClick={openShortcutSettings}
      >
        キーボードショートカット設定
      </button>
    </div>
  ) : (
    <div className="App">
      <span className="App__icon">🦐</span>
      <p className="App__text">evi</p>
      <button
        type="button"
        className="App__settings"
        onClick={openShortcutSettings}
      >
        ⚙️
      </button>
    </div>
  );
};

export default App;
