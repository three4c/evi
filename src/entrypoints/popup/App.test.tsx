import { render } from "@testing-library/react";
import { describe, expect, it } from "vitest";

import App from "./App";

describe("<App />", () => {
  it("Snapshot", () => {
    expect(render(<App />).asFragment()).toMatchSnapshot();
  });
});
