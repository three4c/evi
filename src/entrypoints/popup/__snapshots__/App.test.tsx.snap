// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`<App /> > Snapshot 1`] = `
<DocumentFragment>
  <div
    class="App App--dev"
  >
    <p
      class="App__text"
    >
      Vite Dev Mode
    </p>
    <input
      class="App__child"
      type="text"
    />
    <textarea
      class="App__child"
    >
      aaaaa
bbbbb
ccccc
    </textarea>
    <button
      class="App__child App__button"
      type="button"
    >
      キーボードショートカット設定
    </button>
  </div>
</DocumentFragment>
`;
